worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/clipboard/io/FastSchematicReaderV3.java

```diff
@@ -0,0 +1,326 @@
+import com.sk89q.worldedit.extent.clipboard.BlockArrayClipboard;
+import com.sk89q.worldedit.extent.clipboard.Clipboard;
+import com.sk89q.worldedit.extent.clipboard.io.ClipboardReader;
+import com.sk89q.worldedit.extent.clipboard.io.sponge.VersionedDataFixer;
+import com.sk89q.worldedit.internal.util.LogManagerCompat;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.world.DataFixer;
+import com.sk89q.worldedit.world.biome.BiomeType;
+import com.sk89q.worldedit.world.biome.BiomeTypes;
+import com.sk89q.worldedit.world.block.BlockState;
+import com.sk89q.worldedit.world.block.BlockTypes;
+import com.sk89q.worldedit.world.block.BlockTypesCache;
+import org.apache.logging.log4j.Logger;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.OptionalInt;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.BooleanSupplier;
+import java.util.function.Function;
+
+/**
+ * ClipboardReader for the Sponge Schematic Format v3.
+ * Not necessarily much faster than {@link com.sk89q.worldedit.extent.clipboard.io.sponge.SpongeSchematicV3Reader}, but uses a
+ * stream based approach to keep the memory overhead minimal (especially in larger schematics)
+ */
+@SuppressWarnings("removal") // JNBT
+public class FastSchematicReaderV3 implements ClipboardReader {
+
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private final DataInputStream dataInputStream;
+    private final NBTInputStream nbtInputStream;
+
+    private VersionedDataFixer dataFixer;
+    private MutableBlockVector3 dimensions = MutableBlockVector3.at(0, 0, 0);
+    private BlockVector3 offset;
+    private BlockState[] blockPalette;
+    private BiomeType[] biomePalette;
+    private int dataVersion = -1;
+
+    private boolean blocksWritten = false;
+    private boolean biomesWritten = false;
+    private boolean entitiesWritten = false;
+
+    private boolean needAdditionalIterate = true;
+
+    public FastSchematicReaderV3(
+            DataInputStream dataInputStream,
+            NBTInputStream nbtInputStream
+    ) throws IOException {
+        this.dataInputStream = Objects.requireNonNull(dataInputStream, "dataInputStream");
+        this.nbtInputStream = Objects.requireNonNull(nbtInputStream, "nbtInputStream");
+        if (!dataInputStream.markSupported()) {
+            throw new IOException("InputStream does not support mark");
+        }
+    }
+
+    @Override
+    public Clipboard read(final UUID uuid, final Function<BlockVector3, Clipboard> createOutput) throws IOException {
+        dataInputStream.skipNBytes(1 + 2); // 1 Byte = TAG_Compound, 2 Bytes = Short (Length of tag name = "")
+        dataInputStream.skipNBytes(1 + 2 + 9); // as above + 9 bytes = "Schematic"
+        this.dataInputStream.mark(Integer.MAX_VALUE); // allow resets to basically the start of stream (file) - just skip header
```

Comment:
We should make certain this is a resettableinputstream

Reply:
When used by the BuiltinClipboardFormats enums, the stream is wrapped into a ResettableFileInputStream. Or do you mean, that the variable should show more, that the stream is optimized for resets without flooding the memory?

Reply:
This is public API that we should expect to be used basically